### PR TITLE
published nodes with different "groups/assets" for same endpoint causing closing and reopening of sessions

### PR DESCRIPTION
**Pull request by koepalex**
**Problems:**
* Jobs generated out of published nodes file, all have same id
* GetAvailableJobAsync didn't lock when called from multiple worker
* SendHeartbeatAsync logic with "dirty flag" (_updated) error prone

**Solutions:**
* Use DataGroupId as job id (if available) otherwise include GUID
* Add lock and exception handling to GetAwailableJobAsync
* SendHeartbeatAsync just working on _assignedJobs dictionary

**Tests:**
* added test to verify job ids are unequal

### CODE REVIEW

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs [hunk 1]
```diff
@@ -79,53 +79,79 @@ public class LegacyJobOrchestrator : IJobOrchestrator {
         /// <param name="ct"></param>
         /// <returns></returns>
         public Task<JobProcessingInstructionModel> GetAvailableJobAsync(string workerId, JobRequestModel request, CancellationToken ct = default) {
-            if (_assignedJobs.TryGetValue(workerId, out var job)) {
+            try {
+                _lock.Wait(ct);
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs:83]: Lock should be acquired before the try block. That way, if we fail to acquire it, we will not get to try-catch-finally block and will not release it.

Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs]: good point, updated for GetAvailableJobAsync and SendHearthbeatAsync

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs [hunk 2]
```diff
@@ -79,53 +79,79 @@ public class LegacyJobOrchestrator : IJobOrchestrator {
...
                 return Task.FromResult(job);
             }
-            if (_availableJobs.Count > 0 && _availableJobs.TryDequeue(out job)) {
-                _assignedJobs.AddOrUpdate(workerId, job);
-                if (_availableJobs.Count == 0) {
-                    _updated = false;
-                }
+            catch(OperationCanceledException) {
+                JobProcessingInstructionModel job = null;
+                return Task.FromResult(job);
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs:95]: Why do we return `null` when there was a cancellation and don't propagate the error up ?

Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs]: propagated OperationCanceledExceptions to caller for both SendHeartbeatAsync and GetAvailableJobAsync

**Review by karok2m — APPROVED** [review]

**Pull request merged**